src/com/amazon/corretto/crypto/provider/HkdfRfc5869SecretKeyFactorySpi.java

```diff
@@ -0,0 +1,143 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.corretto.crypto.provider;
+
+import static com.amazon.corretto.crypto.provider.HkdfRfc5869Spec.HKDF_EXTRACT_MODE;
+import static com.amazon.corretto.crypto.provider.HkdfRfc5869Spec.HKDF_MODE;
+
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.HashMap;
+import java.util.Map;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactorySpi;
+import javax.crypto.spec.SecretKeySpec;
+
+class HkdfRfc5869SecretKeyFactorySpi extends SecretKeyFactorySpi {
+  private final int digestCode;
+  private final int digestLength;
+
+  private HkdfRfc5869SecretKeyFactorySpi(final int digestCode, final int digestLength) {
+    this.digestCode = digestCode;
+    this.digestLength = digestLength;
+  }
+
+  @Override
+  protected SecretKey engineGenerateSecret(final KeySpec keySpec) throws InvalidKeySpecException {
+    if (!(keySpec instanceof HkdfRfc5869Spec)) {
+      throw new InvalidKeySpecException("KeySpec must be an instance of HkdfRfc5869Spec");
+    }
+    final HkdfRfc5869Spec spec = (HkdfRfc5869Spec) keySpec;
+
+    final byte[] resultBytes;
+    switch (spec.mode) {
+      case HKDF_MODE:
+        resultBytes = new byte[spec.desiredLength];
+        hkdf(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.salt,
+            spec.salt.length,
+            spec.info,
+            spec.info.length);
+        break;
+      case HKDF_EXTRACT_MODE:
+        // In extract mode, we ignore the value of desiredLength and use
+        // the output size of the digest as desiredLength
+        resultBytes = new byte[digestLength];
+        hkdfExtract(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.salt,
+            spec.salt.length);
+        break;
+      default:
+        // spec.mode is HKDF_EXPAND_MODE. We already checked for it in the constructor of
+        // HkdfRfc5869Spec
```

Comment:
Why not just be explicit and throw on the default case? Comment may be true today, but who knows what junk I'll check-in tomorrow.